akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -38,25 +39,55 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   def flowController: IncomingFlowController = IncomingFlowController.default(settings)
 
-  private var incomingStreams = new immutable.TreeMap[Int, IncomingStreamState]
+  class StreamGroup {
+    // FIXME: this uses 'size' without filtering the incomingStreams collection by state to
+    //  only account for open and halfclosed streams
+    // FIXME: when supporting PUSH_PROMISE this operation should only consider streams
+    //  initiated in this peer
+    def peerHasCapacity: Boolean = {
+      // FIXME: this off-by-one is required because we're evaluating
+      //  capacity after generating demand
+      size < getRemoteMaxConcurrentStreams - 1
+    }
+
+    def getRemoteMaxConcurrentStreams: Int = _maxConcurrentStreams
```

Comment:
Even with PUSH_PROMISE there are only two groups: incoming and outgoing streams which need different behavior. For incoming streams we need to enforce our stream limit, for outgoing streams we need to respect the peer limit. So, it seems packing that logic in here, is not necessary.